### ISSUE TITLE
Preprocessing of data

### ISSUE THREAD

**Issue by jilaniw** [AnkurDeria/MFT#13]
PLz share the code for preprocessing of data, like how the data is normalized and seperated with class indexes? and How patches are created? 
Also when the model output is predicted (as in classification map), is it on the same normalized data used in training?

**Comment by AnkurDeria** [comment]
Here is the base sample code for creating the patches. Please modify it to your use case:

```
def loadData(name):
    data_path = basePath
    if name == 'IP':
        data = sio.loadmat(os.path.join(data_path, 'indian_pines_corrected.mat'))['indian_pines_corrected']
        labels = sio.loadmat(os.path.join(data_path, 'indian_pines_corrected_gt.mat'))['indian_pines_gt']
    elif name == 'SA':
        data = sio.loadmat(os.path.join(data_path, 'Salinas_corrected.mat'))['salinas_corrected']
        labels = sio.loadmat(os.path.join(data_path, 'Salinas_gt.mat'))['salinas_gt']
    elif name == 'UP':
        data = sio.loadmat(os.path.join(data_path, 'paviaU.mat'))['paviaU']
        labels = sio.loadmat(os.path.join(data_path, 'paviaU_gt.mat'))['paviaU_gt']
    
    return data, labels

def splitTrainTestSet(X, y, testRatio, randomState=345):
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=testRatio, random_state=randomState,
                                                        stratify=y)
    return X_train, X_test, y_train, y_test

def applyPCA(X, numComponents=75):
    newX = np.reshape(X, (-1, X.shape[2]))
    pca = PCA(n_components=numComponents, whiten=True)
    newX = pca.fit_transform(newX)
    newX = np.reshape(newX, (X.shape[0],X.shape[1], numComponents))
    return newX, pca

def padWithZeros(X, margin=2):
    newX = np.zeros((X.shape[0] + 2 * margin, X.shape[1] + 2* margin, X.shape[2]))
    x_offset = margin
    y_offset = margin
    newX[x_offset:X.shape[0] + x_offset, y_offset:X.shape[1] + y_offset, :] = X
    return newX

def createImageCubes(X, y, windowSize=5, removeZeroLabels = True):
    margin = int((windowSize - 1) / 2)
    zeroPaddedX = padWithZeros(X, margin=margin)
    # split patches
    patchesData = np.zeros((X.shape[0] * X.shape[1], windowSize, windowSize, X.shape[2]))
    patchesLabels = np.zeros((X.shape[0] * X.shape[1]))
    patchIndex = 0
    for r in range(margin, zeroPaddedX.shape[0] - margin):
        for c in range(margin, zeroPaddedX.shape[1] - margin):
            patch = zeroPaddedX[r - margin:r + margin + 1, c - margin:c + margin + 1]   
            patchesData[patchIndex, :, :, :] = patch
            patchesLabels[patchIndex] = y[r-margin, c-margin]
            patchIndex = patchIndex + 1
    if removeZeroLabels:
        patchesData = patchesData[patchesLabels>0,:,:,:]
        patchesLabels = patchesLabels[patchesLabels>0]
        patchesLabels -= 1
    return patchesData, patchesLabels

def random_unison(a,b, rstate=None):
  assert len(a) == len(b)
  p = np.random.RandomState(seed=rstate).permutation(len(a))
  return a[p], b[p]

X, y = loadData(dataset)
X.shape, y.shape
K = X.shape[2]
K = 20 if dataset == 'IP' else 15
X,pca = applyPCA(X,numComponents=K)
X, y = createImageCubes(X, y, windowSize=windowSize,removeZeroLabels = False)
X.shape, y.shape
if dataset == "IP":
    y_train2 = sio.loadmat(os.path.join('HSI-datasets/Classification/', 'indianpines_disjoint_dset.mat'))\
                                        ['indianpines_disjoint_dset']
    
    ytest = sio.loadmat(os.path.join('HSI-datasets/Classification/', 'indian_pines_corrected_gt.mat'))['indian_pines_gt']
    ytrain = copy.deepcopy(y_train2)
    for i, val in enumerate([0,2,3,5,6,8,10,11,12,14,1,4,7,9,13,15,16]): ytrain[y_train2==i] = val
    del y_train2
    ytest[ytrain!=0] = 0
    
elif dataset == "UP":
    ytrain = sio.loadmat(os.path.join('HSI-datasets/Classification/', 'TRpavia_fixed.mat'))['TRpavia_fixed'].reshape(-1)
    
    ytest = sio.loadmat(os.path.join('HSI-datasets/Classification/', 'TSpavia_fixed.mat'))['TSpavia_fixed'].reshape(-1)
ytrain = ytrain.reshape(-1)
ytest = ytest.reshape(-1)
xtrain = X[ytrain!=0,:]
xtest  = X[ytest!=0,:]
del X
ytrain = ytrain[ytrain!=0] - 1
ytest  = ytest[ytest!=0] - 1
xtrain, ytrain = random_unison(xtrain,ytrain, rstate=None)
```

**Comment by AnkurDeria** [comment]
The classification maps also use the same normalized data including the background pixels. The model is made to predict on the background pixels as well. You can take a look at generate_png function in this [notebook](https://github.com/purbayankar/Multi-Kernel-Rotated-Conv-Deconv-Feature-Fusion-SSRN3d/blob/main/Multi-Kernel-Rotated-Conv-Deconv-Feature-Fusion-SSRN3d.ipynb)
